In [1]:
import pandas as pd
import math
import numpy as np

In [2]:
df = pd.read_csv("Decision_Trees_Training.csv")

In [3]:
df.head()

age  income student credit_rating buys_computer
0     less_than_30    high      no          fair            no
1     less_than_30    high      no     excellent            no
2         31_to_40    high      no          fair           yes
3  greater_than_40  medium      no          fair           yes
4  greater_than_40     low     yes          fair           yes

#### Calculate number of Target Classes and their count

In [39]:
target_coloumn = input("Enter Name of target coloumn : ")
unique , count = np.unique(df[target_coloumn],return_counts=True)
dataset_size = len(df)
print("Unique Labels in Target Class : ",unique)
print("NO and YES count : ",list(count))

Enter Name of target coloumn : buys_computer
Unique Labels in Target Class :  ['no' 'yes']
NO and YES count :  [5, 9]


#### Let's get Entropy of buys_computer coloumn

In [40]:
temp = 0
for x in range(len(unique)):
    temp += -((count[x]/dataset_size) * math.log2(count[x]/dataset_size))
    
print("Entropy of dataset : ",temp)

Entropy of dataset :  0.940285958671


In [41]:
all_coloumns = df.columns.values.tolist()
all_coloumns = all_coloumns[:-1]

#### This function returns Yes and No count in coloumn supplied as parameter.

In [54]:
def get_unique_label_and_count(coloumn_name):
    unique , count = np.unique(df[coloumn_name],return_counts=True)
    print("Unique Labels in {0} Class : {1}".format(coloumn_name,unique))
    print("YES and NO Count : ",list(count))
    return unique , count

In [55]:
def get_Yes_No_count(unique_list,coloumn_name):
    temp_count = []
    for i in range(len(unique_list)):
        yes_count = 0
        no_count = 0
        for j in range(dataset_size):
            if "yes" in df.iloc[j]['buys_computer']:
                if df.iloc[j][coloumn_name] is str(unique_list[i]):
                    yes_count += 1
                    
            elif "no" in df.iloc[j]['buys_computer']:
                if df.iloc[j][coloumn_name] is str(unique_list[i]):
                    no_count += 1
                                    
        temp_count.append(yes_count)
        temp_count.append(no_count)
    
    return temp_count

In [56]:
def calculate_individual_info_coloumn(coloumn_name,result):
    info = []
    for y in range(len(result)-1):
        try:
            total_yes_no = result[y] + result[y+1]
            info.append((result[y]/total_yes_no) * math.log2((result[y]/total_yes_no)) + (result[y+1]/total_yes_no) * math.log2((result[y+1]/total_yes_no)))

        except ValueError:
            info.append(0)
            continue
                    
    info = info[0::2]
    info = list(map(abs,info))
    print("Individual Info of {0} is {1} : ".format(coloumn_name,info))
    return info

In [61]:
def calculate_info_coloumn(col_name,count,ind_col_info):
    temp_info = 0
    print(ind_col_info)
    for x in range(len(count)):
        temp_info += (count[x]/dataset_size) * ind_col_info[x]
    
    print("Info of {0} is : {1}".format(col_name,temp_info))
    print("\n")
    return temp_info

In [62]:
final_info_list = []
for i in range(len(all_coloumns)):
    global final_info_list
    count = get_unique_label_and_count(all_coloumns[i])
    
    result = get_Yes_No_count(count[0],all_coloumns[i])
    
    ind_col = calculate_individual_info_coloumn(all_coloumns[i],result)
    
    info = calculate_info_coloumn(all_coloumns[i],count[1],ind_col)
    
    final_info_list.append(info)

Unique Labels in age Class : ['31_to_40' 'greater_than_40' 'less_than_30']
YES and NO Count :  [4, 5, 5]
Individual Info of age is [0, 0.9709505944546686, 0.9709505944546686] : 
[0, 0.9709505944546686, 0.9709505944546686]
Info of age is : 0.6935361388961918


Unique Labels in income Class : ['high' 'low' 'medium']
YES and NO Count :  [4, 4, 6]
Individual Info of income is [1.0, 0.8112781244591328, 0.9182958340544896] : 
[1.0, 0.8112781244591328, 0.9182958340544896]
Info of income is : 0.9110633930116763


Unique Labels in student Class : ['no' 'yes']
YES and NO Count :  [7, 7]
Individual Info of student is [0.9852281360342516, 0.5916727785823275] : 
[0.9852281360342516, 0.5916727785823275]
Info of student is : 0.7884504573082896


Unique Labels in credit_rating Class : ['excellent' 'fair']
YES and NO Count :  [6, 8]
Individual Info of credit_rating is [1.0, 0.8112781244591328] : 
[1.0, 0.8112781244591328]
Info of credit_rating is : 0.8921589282623617




In [63]:
final_gain_list = []
for x in range(len(final_info_list)):
    final_gain_list.append(temp - final_info_list[x])

print("Max Gain : ",max(final_gain_list))

Max Gain :  0.246749819774


## Conclusion : 0.2467 is maximum gain obtained.
### So, age attribute serves as ROOT of our decision tree.